Modules/_winapi.c

```diff
@@ -774,12 +774,139 @@ gethandle(PyObject* obj, const char* name)
     return ret;
 }
 
+static PyObject*
+sortenvironmentkey(PyObject *module, PyObject *item)
+{
+    return _winapi_LCMapStringEx_impl(NULL, LOCALE_NAME_INVARIANT, LCMAP_UPPERCASE, item);
+}
+
+static PyMethodDef sortenvironmentkey_def = {
+    "sortenvironmentkey",
+    _PyCFunction_CAST(sortenvironmentkey),
+    METH_O,
+    "",
+};
+
+static PyObject *
+normalize_environment(PyObject* environment)
+{
+    PyObject *result = NULL, *keys = NULL, *keyfunc = NULL, *kwnames = NULL;
+
+    keys = PyMapping_Keys(environment);
+    if (keys == NULL) {
+        return NULL;
+    }
+
+    keyfunc = PyCFunction_New(&sortenvironmentkey_def, NULL);
+    if (keyfunc == NULL) {
+        goto error;
+    }
+    kwnames = PyTuple_New(1);
+    if (kwnames == NULL) {
+        goto error;
+    }
+    assert(PyTuple_SetItem(kwnames, 0, PyUnicode_FromString("key")) == 0);
+    PyObject *args[] = { keys, keyfunc };
+    if (PyObject_VectorcallMethod(&_Py_ID(sort), args, 1, kwnames) == NULL) {
+        goto error;
+    }
+
+    result = PyDict_New();
+    if (result == NULL) {
+        goto error;
+    }
+    wchar_t *prev_key_string = NULL;
+
+    for (int i=0; i<PyList_GET_SIZE(keys); i++) {
+        PyObject *key = PyList_GET_ITEM(keys, i);
+        if (key == NULL) {
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
```

Comment:
```suggestion
            goto error;
```

Just saw the discussion here about `PyErr_Occurred`, and while that _should_ work (it's going to be checked on return and a `SystemError` will be raised if you return `NULL` without an exception set), the little `goto` dance I have in my suggestion is fine.